enhancements/network/bgp-ovn-kubernetes.md

```diff
@@ -0,0 +1,685 @@
+the proper VRF, where it will be forwarded via ovn-k8s-mp0 of the respective user-defined network. Today when these
+packets ingress ovn-k8s-mp0 into OVN, they are SNAT'ed to the ovn-k8s-mp0 address (.2 address). Reasons for this include:
+1. To ensure incoming traffic that may be routed to another node via geneve would return back to this node. This is undesirable
+and unnecessary for networks where only their respective per-node pod subnet is being advertised to the BGP fabric.
+2. If running shared gateway mode, the reply packet would be routed via br-ex with the default gateway configured in
+ovn_cluster_router.
+
+For local gateway mode, changes will need to be made to skip the SNAT as it is not necessary and provides pods with the
+true source IP of the sender.
+
+
+Additionally, modifications will need to be made in the kernel routing table to
+leak routes to the pod subnets from each user-defined VRF into the default VRF routing table.
+
+#### Deployment Considerations
+
+##### MetalLB
+
+MetalLB is supported today as a day 2 add-on operator. MetalLB supports FRR mode, and has recently added support for
+FRR-K8S mode. When the MetalLB Operator deploys MetalLB in FRR-K8S mode, it handles instantiating FRR-K8S. For the purposes
+of OVN-Kubernetes integration, we need this as a day 0 function. For pods relying on the underlay and not using an overlay
+encapsulation, the BGP routes will need to be learned by the time the pods come up.
+
+In order to achieve day 0 functionality, FRR-K8S will be launched by CNO as a host networked pod on each node (where
+applicable). MetalLB Operator will be modified so that it can launch in FRR-K8S mode, without deploying FRR-K8S, and
+integrate with the one launched by CNO.
+
+This means, there will be two ways of operating:
+- The current behaviour, which is: CNO does not deploy FRR-K8s, the MetalLB Operator deploys both MetalLB and FRR-K8s
+- CNO deploys FRR-K8s and MetalLB leverages the FRR-K8s instance deployed by CNO
+
+###### MetalLB deploys FRR-K8s
+This is the behaviour implemented in 4.16 as Tech Preview and planned to be GA in 4.17. The users don't care about pod
+to pod reachability with BGP, they want to use MetalLB and possibly FRR-K8s for learning routes on the nodes.
+The MetalLB operator will deploy FRR-K8s and MetalLB in the same namespace.
+
+###### CNO deploys FRR-K8s
+In this version, CNO deploys FRR-K8s in a separate namespace, so it can be used for the extra features described in this
+proposal. The user will need to set:
+
+- A flag on the CNO configuration, to deploy FRR-K8s and to instruct OVN-K of the presence of the FRR controller
+- A deployment mode "FRR-K8s already present" in the "MetalLB" resource, to instruct the operator to deploy MetalLB in
+FRR-K8s mode but without deploying FRR-K8s
+- In this scenario the permissions given to MetalLB over the FRR-K8s resources must be changed from namespace scoped to
```

Comment:
let's discuss this one over slack as well.

Reply:
spoke to @oribon , there's also another item to consider / to implement when we separate the namespaces, which is:

Both MetalLB and FRR-K8s handle the passwords of the session as a secret ref. The workflow is:

- the user writes the secret in the metallb ns
- the user writes the bgppeer with a reference to that secret (as a secretRef field)
- metallb generates an FRRConfiguration referencing that secret

With two separate namespaces, this won't be possible anymore so what MetalLB will have to do is, get permissions to write on the frr-k8s namespace and mirror the secrets (we don't want to have frr-k8s to read the one from the metallb's namespace as that is not know upfront, being it a day 2 operator, nor we want to give it permissions to all the secrets of the cluster).

Reply:
We decided CNO will always deploy FRR-K8S and register the CRDs when a new OpenShift API is configured to enable FRR-K8S. When MetalLB is launched, it will enable this same API in order to provoke CNO to launch FRR-K8S.

@fedepaol re: your comments about secrets. If a user was not using metallb, but wanted to use secrets, they should just create the secret in the frr-k8s namespace directly right?

Reply:
> We decided CNO will always deploy FRR-K8S and register the CRDs when a new OpenShift API is configured to enable FRR-K8S. When MetalLB is launched, it will enable this same API in order to provoke CNO to launch FRR-K8S.
> 
> @fedepaol re: your comments about secrets. If a user was not using metallb, but wanted to use secrets, they should just create the secret in the frr-k8s namespace directly right?

Right